apps/extension/src/shared/components/grpc-endpoint-form/use-grpc-endpoint-form.ts

```diff
@@ -0,0 +1,117 @@
+
+const isValidUrl = (url: string) => {
+  try {
+    new URL(url);
+    return true;
+  } catch {
+    return false;
+  }
+};
+
+export const useGrpcEndpointForm = () => {
+  const [originalChainId, setOriginalChainId] = useState<string | undefined>();
+  const [chainId, setChainId] = useState<string>();
+  const grpcEndpoints = useMemo(() => [...GRPC_ENDPOINTS].sort(randomSort), []);
+  const { grpcEndpoint, setGrpcEndpoint } = useStoreShallow(useSaveGrpcEndpointSelector);
+  const [grpcEndpointInput, setGrpcEndpointInput] = useState<string>(
+    grpcEndpoint ?? grpcEndpoints[0]?.url ?? '',
+  );
+  const [rpcError, setRpcError] = useState<string>();
+  const [isSubmitButtonEnabled, setIsSubmitButtonEnabled] = useState(false);
+  const isCustomGrpcEndpoint = !GRPC_ENDPOINTS.some(({ url }) => url === grpcEndpointInput);
+
+  const setGrpcEndpointInputOnLoadFromState = useCallback(() => {
+    if (grpcEndpoint) setGrpcEndpointInput(grpcEndpoint);
+  }, [grpcEndpoint]);
+
+  useEffect(setGrpcEndpointInputOnLoadFromState, [setGrpcEndpointInputOnLoadFromState]);
+
+  const handleChangeGrpcEndpointInput = useMemo(() => {
+    return debounce(async (grpcEndpointInput: string) => {
+      setIsSubmitButtonEnabled(false);
+      setRpcError(undefined);
+
+      if (!isValidUrl(grpcEndpointInput)) return;
+
+      try {
+        const trialClient = createPromiseClient(
+          QueryService,
+          createGrpcWebTransport({ baseUrl: grpcEndpointInput }),
+        );
+        const { appParameters } = await trialClient.appParameters({});
+        if (!appParameters?.chainId) throw new ConnectError('', Code.NotFound);
+
+        setIsSubmitButtonEnabled(true);
+        setChainId(appParameters.chainId);
+
+        // Only set the original chain ID the first time, so that we can compare
+        // it on submit.
+        setOriginalChainId(originalChainId =>
+          originalChainId ? originalChainId : appParameters.chainId,
+        );
+      } catch (e) {
+        if (e instanceof ConnectError && e.code === Code.NotFound) {
+          setRpcError(
+            'Could not get a chain ID from this endpoint. Please double-check your endpoint URL and try again.',
+          );
+        } else if (e instanceof ConnectError && e.code === Code.Unknown) {
+          setRpcError(
+            'Could not connect to endpoint. Please double-check your endpoint URL and try again.',
+          );
+        } else {
+          setRpcError('Could not connect to endpoint: ' + String(e));
+        }
+      }
+    }, 400);
+  }, []);
+
+  useEffect(
+    () => void handleChangeGrpcEndpointInput(grpcEndpointInput),
+    [handleChangeGrpcEndpointInput, grpcEndpointInput],
+  );
+
+  const onSubmit = async (
+    /** Callback to run when the RPC endpoint successfully saves */
+    onSuccess: () => void | Promise<void>,
+  ) => {
+    setIsSubmitButtonEnabled(false);
+    await setGrpcEndpoint(grpcEndpointInput);
+    // If the chain id has changed, our cache is invalid
```

Comment:
on mainnet, this is a more serious condition than is typical in our experience as developers constantly flipping between testnet/preview. there should be a confirmation or other significant ui hurdle at this point.

Reply:
On it

Reply:
OK, I added a confirmation dialog in [this commit](https://github.com/penumbra-zone/web/pull/877/commits/4251dd970f81c78a9e347f51d1519f867b1e6954?diff=split&w=1).

Was the rest of this PR a ✅ ? If so, could you please approve if that latest commit looks good?